### PR TITLE
[YOLOV8] `ultralytics` Update

### PR DESCRIPTION
**Pull request by dsikka**
### Summary
- Update sparseml to use `ultralytics` == 8.0.124. 
- Also update some of the workflows which did not work currently (such as exporting segmentation models with samples)

### Testing
- Training, Validation, and Export workflows were tested for a dense network and while applying a recipe during training
- Recipe applied during training:

```
version: 1.1.0

training_modifiers:
  - !EpochRangeModifier
    start_epoch: 0
    end_epoch: 15

pruning_modifiers:
  - !GMPruningModifier
    init_sparsity: 0.05
    final_sparsity: 0.10
    start_epoch: 5.0
    end_epoch: 10.0
    update_frequency: 1.0
    params: ["model.0.conv.weight", "model.1.conv.weight"]

quantization_modifiers:
  - !QuantizationModifier
    start_epoch: 11.0
    freeze_bn_stats_epoch: 12.0
    disable_quantization_observer_epoch: 13.0
    ignore: ["Upsample", "Concat", "SiLU"]
```

#### Testing Example Commands:

GPU + Recipe

```
Train: sparseml.ultralytics.train --recipe ~/sparseml/recipe.yaml (
Export: sparseml.ultralytics.export_onnx --model runs/detect/train/weights/last.pt --export-samples 20
Val: sparseml.ultralytics.val --model runs/detect/train/weights/[best.pt](http://best.pt/)

```
GPU + Dense

```
Train: sparseml.ultralytics.train  --epochs 50
Export: sparseml.ultralytics.export_onnx --model runs/detect/train/weights/last.pt --export-samples 20
Val: parseml.ultralytics.val --model runs/detect/train/weights/best.pt

```
GPU + Recipe - Segmentation
```
Train: sparseml.ultralytics.train --recipe ~/sparseml/recipe.yaml --task segment --model yolov8n-seg.yaml --data coco128-seg.yaml
Export: sparseml.ultralytics.export_onnx --model runs/segment/train/weights/last.pt --export-samples 20
Val: sparseml.ultralytics.val --task segment --data coco128-seg.yaml --model runs/segment/train/weights/last.pt
```

GPU + Dense - Segmentation
```
Train: sparseml.ultralytics.train --epochs 50 --task segment --model yolov8n-seg.yaml --data coco128-seg.yaml
Export: sparseml.ultralytics.export_onnx --model runs/segment/train/weights/last.pt --export-samples 20
Val: sparseml.ultralytics.val --task segment --data coco128-seg.yaml --model runs/segment/train/weights/last.pt
```

- Trained + exported models were visually examined using netron to make sure the recipes were applied properly during training
- Validation loss was compared for quantized models before and after applying the changes to make sure they were similar

Ran validation on the following model: 
```
sparseml.ultralytics.val \
  --model zoo:cv/detection/yolov8-s/pytorch/ultralytics/coco/pruned65_quant-none \
  --data coco.yaml
```

The mAP@0.50 and mAP@0.5:0.95 match the values listed on sparsezoo and the predictions plots also make sense

### CODE REVIEW

**Comment by abhinavnmagic** [comment]
Apart from the validation loss, were the mAP scores also similar?
Can we also do a run including all the parameters for pruning/quantization? Additionally, utilizing the schema overrides for quantization recipe. A good way to check would be to pick a released [recipe](https://sparsezoo.neuralmagic.com/models/yolov8-s-coco-pruned65_quantized?hardware=deepsparse-c6i.12xlarge&comparison=yolov8-s-coco-base&tab=3) from sparsezoo and running that for a few number of epochs. This would help us verify if the changes are applied as expected to all the layers in the network, and correct recipes if needed.

File: src/sparseml/yolov8/trainers.py [hunk 1]
```diff
@@ -581,7 +598,7 @@ def __init__(self, model="yolov8n.yaml", type="v8") -> None:
         elif self.ValidatorClass == SegmentationValidator:
             self.ValidatorClass = SparseSegmentationValidator
 
-    def _load(self, weights: str):
+    def _load(self, weights: str, task=None):
```
Comment (abhinavnmagic) [inline src/sparseml/yolov8/trainers.py:601]: Since we are modifying the _load function, we should check if we get the same accuracy when running validation on sparsezoo-published models.

Comment (dsikka) [reply src/sparseml/yolov8/trainers.py:601]: The `_load` function wasn't really changed apart from now taking an input task and using that if the yolov8 itself can't guess the task from the loaded model.  Do you have a suggested model that you'd want me to run?

**Comment by dsikka** [comment]
> Apart from the validation loss, were the mAP scores also similar? Can we also do a run including all the parameters for pruning/quantization? Additionally, utilizing the schema overrides for quantization recipe. A good way to check would be to pick a released [recipe](https://sparsezoo.neuralmagic.com/models/yolov8-s-coco-pruned65_quantized?hardware=deepsparse-c6i.12xlarge&comparison=yolov8-s-coco-base&tab=3) from sparsezoo and running that for a few number of epochs. This would help us verify if the changes are applied as expected to all the layers in the network, and correct recipes if needed.

Just so I understand, you're suggesting training using one of the released recipes?

File: src/sparseml/yolov8/trainers.py (hunk 1 above)
Comment (abhinavnmagic) [reply src/sparseml/yolov8/trainers.py:601]: Any sparseml obtained yolov8 should do. You could use [model](zoo:cv/detection/yolov8-s/pytorch/ultralytics/coco/pruned65_quant-none). I don't expect to see an error, but erring on the side of caution.

**Comment by abhinavnmagic** [comment]
> > Apart from the validation loss, were the mAP scores also similar? Can we also do a run including all the parameters for pruning/quantization? Additionally, utilizing the schema overrides for quantization recipe. A good way to check would be to pick a released [recipe](https://sparsezoo.neuralmagic.com/models/yolov8-s-coco-pruned65_quantized?hardware=deepsparse-c6i.12xlarge&comparison=yolov8-s-coco-base&tab=3) from sparsezoo and running that for a few number of epochs. This would help us verify if the changes are applied as expected to all the layers in the network, and correct recipes if needed.
> 
> Just so I understand, you're suggesting training using one of the released recipes?

Yes, exactly.

File: src/sparseml/yolov8/trainers.py (hunk 1 above)
Comment (dsikka) [reply src/sparseml/yolov8/trainers.py:601]: Ran validation on the following model: 
```
sparseml.ultralytics.val \
  --model zoo:cv/detection/yolov8-s/pytorch/ultralytics/coco/pruned65_quant-none \
  --data coco.yaml
```

The mAP@0.50 and mAP@0.5:0.95 match the values listed on sparsezoo and the predictions also make sense

File: src/sparseml/yolov8/utils/export_samples.py [hunk 2]
```diff
@@ -147,10 +147,20 @@ def _export_torch_outputs(
...
+        pred_0 = preds[0].detach().to("cpu")
+        pred_1 = preds[1].detach().to("cpu")
 
-    sample_output_filename = os.path.join(sample_out_dir, f"out-{file_idx}.npz")
-    numpy.savez(sample_output_filename, preds)
+        sample_output_filename = os.path.join(sample_out_dir, f"out-{file_idx}-0.npz")
+        numpy.savez(sample_output_filename, pred_0)
+
+        sample_output_filename = os.path.join(sample_out_dir, f"out-{file_idx}-1.npz")
+        numpy.savez(sample_output_filename, pred_1)
```
Comment (bfineran) [inline src/sparseml/yolov8/utils/export_samples.py:159]: let's keep both to the same file - npz supports holding multiple arrays.  you can pass them as explicit kwarg names

**Review by bfineran — DISMISSED** [review]
LGTM pending comment

**Review by abhinavnmagic — APPROVED** [review]

**Review by abhinavnmagic — APPROVED** [review]

**Review by bfineran — APPROVED** [review]
